feat: Implement save / restore viewer state (#1523)
cognitedata/reveal: documentation/docs/examples/cad-save-viewerstate.mdx

      },
      "modelId": 3356984403684032,
      "revisionId": 6664823881595566,
      "styledSets": [
        {
          "token": "ByNodePropertyNodeSet",
          "state": {
            "PDMS": {
              "Type": "PIPE"
            }
          },
          "options": {
            "requestPartitions": 1
          },
          "appearance": {
            "renderGhosted": false,
            "color": [
              200,
              40,
              20
            ]
          }
        }
      ]
    }
  ]
};

viewer.setViewState(myState);
```

## Serializing and deserializing custom node sets
When working with custom node sets, the Reveal viewer needs to know how to serialize and deserialize the set.
Serialization is handled through the abstract method `serialize` which needs to be implemented when extending `NodeSet`.
The output of `serialize` is of type `SerializedNodeSet` and contains a `token` which is used to recreate the `NodeSet` during deserialization, and a `state` object which contains the data needed to restore the `NodeSet`.
It is up to the application developer to fill this state object with any nessesary data needed to recreate the state of the `NodeSet`.

> And optional `options` object can also be supplied containing any metadata for loading data (e.g. the number of API calls to fetch the needed data in parallel).

Reviewer: ```suggestion
An optional `options` object can also be supplied containing any metadata for loading data (e.g. the number of API calls to fetch the needed data in parallel).

```

> If contents returned from `serialize` is contained in the result of `getViewState`.

Reviewer: If... then (?). Didn't get this

In addition to serialization, Reveal also needs to know how to recreate the custom `NodeSet` during deserialization.
This is done by calling the free-function `registerCustomNodeSetType` with the same `token` defined by the custom `NodeSet` and a lambda function which instantiates and restore the state of the `NodeSet`.

> The lambda function has two parameters: the `descriptor` containing the state and options which should be applied to the `NodeSet` and the context containing an instance of the [CogniteSDK](https://docs.cognite.com/dev/guides/sdk/js/) and the `Cognite3DModel` which the `NodeSet` was originally applied to.

Reviewer: ```suggestion
The lambda function has two parameters: the `descriptor` containing the state and options which should be applied to the `NodeSet` and the context containing an instance of the [CogniteClient from the Cognite SDK](https://docs.cognite.com/dev/guides/sdk/js/) and the `Cognite3DModel` which the `NodeSet` was originally applied to.

```

It is up to the application developer how and if they want to use the context and/or descriptor, but it is provided as a convinience. 

> And example of a simple custom `NodeSet` can be seen below.

Reviewer: ```suggestion
An example of a simple custom `NodeSet` can be seen below.

```

For more information on creating a custom `NodeSet`, see [Create custom node sets](cad-styling-custom).
```jsx

> // MyNodeSet.ts
> class MyNodeSet extends NodeSet {

Reviewer: ```suggestion
// MyNodeSet.ts

import { NodeSet, registerCustomNodeSetType } from '@cognite/reveal';

class MyNodeSet extends NodeSet {

```

In the example below, a previously stored state is applied which applies a red color to nodes that have an attribute `'PDMS.Type'`.

:::note
Setting the viewer state assumes that model has already been loaded. 
An error will be thrown if a given state is loaded for a model which does not exist.  
:::

```js runnable

const myState = {
  "camera": {
    "position": {
      "x": 165.603,
      "y": 41.649,
      "z": -10.940
    },
    "target": {
      "x": 130.537,
      "y": 49.138,
      "z": -119.095
    }
  },
  "models": [
    {
      "defaultNodeAppearance": {
        "renderGhosted": true
      },
      "modelId": 3356984403684032,
      "revisionId": 6664823881595566,
      "styledSets": [
        {
          "token": "ByNodePropertyNodeSet",
          "state": {
            "PDMS": {
              "Type": "PIPE"
            }
          },
          "options": {
            "requestPartitions": 1
          },
          "appearance": {
            "renderGhosted": false,
            "color": [
              200,
              40,
              20
            ]
          }
        }
      ]
    }
  ]
};

viewer.setViewState(myState);

> ```

Author: Ah, thanks! Seems it was due to minifing, cause the class.name to be wrong. Should be fixed with static names.
